std/fmt/parse_float.zig

```diff
@@ -0,0 +1,425 @@
+        w += u64(d.d1) -% u64(s.d1);
+        d.d1 = @truncate(u32, w);
+
+        w >>= 32;
+        w += u64(d.d2) -% u64(s.d2);
+        d.d2 = @truncate(u32, w);
+    }
+
+    fn dump(d: Z96) void {
+        std.debug.warn("{} {} {}\n", d.d0, d.d1, d.d2);
+    }
+};
+
+const FloatRepr = struct {
+    negative: bool,
+    exponent: i32,
+    mantissa: u64,
+};
+
+fn convertRepr(comptime T: type, n: FloatRepr) T {
+    const mask28: u32 = 0xf << 28;
+
+    var s: Z96 = undefined;
+    var q: Z96 = undefined;
+    var r: Z96 = undefined;
+
+    s.d0 = @truncate(u32, n.mantissa);
+    s.d1 = @truncate(u32, n.mantissa >> 32);
+    s.d2 = 0;
+
+    var binary_exponent: u64 = 92;
+    var exp = n.exponent;
+
+    while (exp > 0) : (exp -= 1) {
+        q.shiftLeft1(s); // q = p << 1
+        r.shiftLeft1(q); // r = p << 2
+        s.shiftLeft1(r); // p = p << 3
+        q.add(s); // p = (p << 3) + (p << 1)
+
+        exp -= 1;
+
+        while (s.d2 & mask28 != 0) {
+            q.shiftRight1(s);
+            binary_exponent += 1;
+            s = q;
+        }
+    }
+
+    while (exp < 0) {
+        while (s.d2 & (1 << 31) == 0) {
+            q.shiftLeft1(s);
+            binary_exponent -= 1;
+            s = q;
+        }
+
+        q.d2 = s.d2 / 10;
+        r.d1 = s.d2 % 10;
+        r.d2 = (s.d1 >> 8) | (r.d1 << 24);
+        q.d1 = r.d2 / 10;
+        r.d1 = r.d2 % 10;
+        r.d2 = ((s.d1 & 0xff) << 16) | (s.d0 >> 16) | (r.d1 << 24);
+        r.d0 = r.d2 / 10;
+        r.d1 = r.d2 % 10;
+        q.d1 = (q.d1 << 8) | ((r.d0 & 0x00ff0000) >> 16);
+        q.d0 = r.d0 << 16;
+        r.d2 = (s.d0 *% 0xffff) | (r.d1 << 16);
+        q.d0 |= r.d2 / 10;
+        s = q;
+
+        exp += 1;
+    }
+
+    if (s.d0 != 0 or s.d1 != 0 or s.d2 != 0) {
+        while (s.d2 & mask28 == 0) {
+            q.shiftLeft1(s);
+            binary_exponent -= 1;
+            s = q;
+        }
+    }
+
+    binary_exponent += 1023;
+
+    const repr: u64 = blk: {
+        if (binary_exponent > 2046) {
+            break :blk if (n.negative) f64_minus_infinity else f64_plus_infinity;
+        } else if (binary_exponent < 1) {
+            break :blk if (n.negative) f64_minus_zero else f64_plus_zero;
+        } else if (s.d2 != 0) {
+            const binexs2 = u64(binary_exponent) << 52;
+            const rr = (u64(s.d2 & ~mask28) << 24) | ((u64(s.d1) + 128) >> 8) | binexs2;
+            break :blk if (n.negative) rr | (1 << 63) else rr;
+        } else {
+            break :blk 0;
+        }
+    };
+
+    const f = @bitCast(f64, repr);
+    return @floatCast(T, f);
+}
+
+const State = enum {
+    SkipLeadingWhitespace,
+    MaybeSign,
+    LeadingMantissaZeros,
+    LeadingFractionalZeros,
+    MantissaIntegral,
+    MantissaFractional,
+    ExponentSign,
+    LeadingExponentZeros,
+    Exponent,
+    Stop,
+};
+
+const ParseResult = enum {
+    Ok,
+    PlusZero,
+    MinusZero,
+    PlusInf,
+    MinusInf,
+};
+
+inline fn isDigit(c: u8) bool {
```

Comment:
Would an `ascii` module be valuable? Similar to `<ctype.h>`. There are some implementations under `os/path.zig` that perform similar functions.

Reply:
I think that sounds quite reasonable.